File: apps/native/modules/remote-identity-custody/ios/RemoteIdentityCustodyModule.swift
import ExpoModulesCore
import Foundation
import Security

// RemoteIdentityCustody — iOS durable P-256 custody.
//
// A durable handle is a non-exportable P-256 private key held in the Secure
// Enclave (enclave profile) or as a Keychain SecKey (software-backed profile).
// The `handleId` returned to JS is the key's random application tag
// (`kSecAttrApplicationTag`). JS never receives private key bytes.
//
// Generation metadata (the app-layer monotonic high-water mark that maps a tag
// to its generation) is persisted separately as a non-synchronizable,
// `ThisDeviceOnly` Keychain generic-password item — see the app-layer
// `NativeCustodyStore`. The private key itself is created with
// `kSecAttrAccessibleWhenUnlockedThisDeviceOnly` and is never marked
// synchronizable, so it can neither leave the device nor sync to iCloud.
//
// Diffie-Hellman key agreement lives entirely outside this module (the Noise
// binding owns it); this module is P-256 signing only.
public class RemoteIdentityCustodyModule: Module {
  private let keyType = kSecAttrKeyTypeECSECPrimeRandom
  private let p256Order = P256Scalar.order
  private let p256HalfOrder = P256Scalar.halfOrder

  // MARK: - Native platform backing (UNIMPLEMENTED — fails closed)

  // TODO(native-platform): the durable native custody backing is NOT wired yet.
  // Until it is, every entry point fails closed rather than returning a
  // plausible-but-wrong result or relying on volatile state. What remains:
  //   1. a real durable store for the tag->generation high-water mark
  //      (non-synchronizable ThisDeviceOnly Keychain generic-password item), and
  //   2. recovery of a key's real profile/presence from its SecAccessControl /
  //      accessibility attributes (see profileForTag below).
  // The reference implementation of each method is retained below to document
  // the intended Keychain / Secure Enclave mechanism. This MUST be built and
  // verified on the iOS CI leg — it cannot be compiled on the Linux gate box.
  // See NATIVE-PLATFORM-TODO.md.
  private func requireNativeBackingWired() throws {
    throw CustodyException(
      "unimplemented on this platform: durable native custody backing is not wired — failing closed (see NATIVE-PLATFORM-TODO.md)"
    )
  }

  public func definition() -> ModuleDefinition {
    Name("RemoteIdentityCustody")

    AsyncFunction("generateP256") { (handleId: Data, profile: String, requireUserPresence: Bool) -> [String: Any] in
      try self.generate(handleId: handleId, profile: profile, requireUserPresence: requireUserPresence)
    }

    AsyncFunction("signP256") { (handleId: Data, signingMessage: Data) -> Data in
      try self.sign(handleId: handleId, message: signingMessage)
    }

    AsyncFunction("publicKey") { (handleId: Data) -> [String: Any] in
      try self.publicKeyReport(handleId: handleId)
    }

    AsyncFunction("rotateP256") { (handleId: Data, newHandleId: Data) -> [String: Any] in
      try self.rotate(handleId: handleId, newHandleId: newHandleId)
    }

    AsyncFunction("destroyGeneration") { (handleId: Data) -> Void in
      try self.destroy(handleId: handleId)
    }
  }

  // MARK: - Generation

  private func generate(handleId: Data, profile: String, requireUserPresence: Bool) throws -> [String: Any] {
    try requireNativeBackingWired() // TODO(native-platform): fails closed until wired.
    // The caller-assigned handle IS the keystore application tag, so a durable
    // write-ahead marker can name the key before it exists.
    let tag = handleId
    let inEnclave = profile == "ios-secure-enclave"

    var keyAttributes: [String: Any] = [
      kSecAttrKeyType as String: keyType,
      kSecAttrKeySizeInBits as String: 256,
      kSecPrivateKeyAttrs as String: [
        kSecAttrIsPermanent as String: true,
        kSecAttrApplicationTag as String: tag,
      ],
    ]

    // The durable key is device-bound and non-exportable: ThisDeviceOnly
    // accessibility, never synchronizable. Presence profiles additionally
    // require a live user via `.userPresence`.
    let flags: SecAccessControlCreateFlags =
      requireUserPresence ? [.privateKeyUsage, .userPresence] : [.privateKeyUsage]
    guard
      let access = SecAccessControlCreateWithFlags(
        kCFAllocatorDefault,
        kSecAttrAccessibleWhenUnlockedThisDeviceOnly,
        flags,
        nil
      )
    else {
      throw CustodyException("failed to create access control")
    }
    var privateAttrs = keyAttributes[kSecPrivateKeyAttrs as String] as! [String: Any]
    privateAttrs[kSecAttrAccessControl as String] = access
    keyAttributes[kSecPrivateKeyAttrs as String] = privateAttrs

    if inEnclave {
      // Secure Enclave residency: keys are generated inside the SEP and cannot
      // be extracted. `kSecAttrTokenIDSecureEnclave` pins residency.
      keyAttributes[kSecAttrTokenID as String] = kSecAttrTokenIDSecureEnclave
    }

    var error: Unmanaged<CFError>?
    guard let privateKey = SecKeyCreateRandomKey(keyAttributes as CFDictionary, &error) else {
      throw CustodyException("SecKeyCreateRandomKey failed: \(Self.describe(error))")
    }

    let (x, y) = try Self.publicCoordinates(of: privateKey)
    let attestation = try Self.attest(privateKey: privateKey, profile: profile, requireUserPresence: requireUserPresence)
    let evidence = Self.providerEvidence(tag: tag, attestation: attestation)

    return [
      "handleId": tag,
      "publicKey": ["x": x, "y": y],
      "attestation": attestation,
      "providerEvidence": evidence,
    ]
  }

  // MARK: - Rotation

  private func rotate(handleId: Data, newHandleId: Data) throws -> [String: Any] {
    try requireNativeBackingWired() // TODO(native-platform): fails closed until wired.
    // The old key must exist; the new key is created under the caller-assigned
    // `newHandleId` tag and the old key is retained until destroyed after publish.
    _ = try loadPrivateKey(tag: handleId)
    let existing = try Self.profileForTag(handleId)
    return try generate(
      handleId: newHandleId,
      profile: existing.profile,
      requireUserPresence: existing.requireUserPresence
    )
  }

  // MARK: - Signing

  private func sign(handleId: Data, message: Data) throws -> Data {
    try requireNativeBackingWired() // TODO(native-platform): fails closed until wired.
    let privateKey = try loadPrivateKey(tag: handleId)
    var error: Unmanaged<CFError>?
    // Message-based signing: the platform hashes with SHA-256 internally.
    guard
      let derSignature = SecKeyCreateSignature(
        privateKey,
        .ecdsaSignatureMessageX962SHA256,
        message as CFData,
        &error
      ) as Data?
    else {
      throw CustodyException("SecKeyCreateSignature failed: \(Self.describe(error))")
    }
    // Convert DER (SEQUENCE { INTEGER r, INTEGER s }) to fixed 64-byte P1363 and
    // normalize to low-S. Zero / out-of-range components are corruption, not
    // normalized away.
    return try P256Scalar.derToLowSP1363(derSignature)
  }

  // MARK: - Public key

  private func publicKeyReport(handleId: Data) throws -> [String: Any] {
    try requireNativeBackingWired() // TODO(native-platform): fails closed until wired.
    let privateKey = try loadPrivateKey(tag: handleId)
    let (x, y) = try Self.publicCoordinates(of: privateKey)
    let existing = try Self.profileForTag(handleId)
    let attestation = try Self.attest(
      privateKey: privateKey,
      profile: existing.profile,
      requireUserPresence: existing.requireUserPresence
    )
    return ["x": x, "y": y, "attestation": attestation]
  }

  // MARK: - Destroy

  private func destroy(handleId: Data) throws {
    try requireNativeBackingWired() // TODO(native-platform): fails closed until wired.
    let query: [String: Any] = [
      kSecClass as String: kSecClassKey,
      kSecAttrApplicationTag as String: handleId,
      kSecAttrKeyType as String: keyType,
    ]
    let status = SecItemDelete(query as CFDictionary)
    if status != errSecSuccess && status != errSecItemNotFound {
      throw CustodyException("SecItemDelete failed: \(status)")
    }
    if status == errSecItemNotFound {
      throw CustodyException("handle not found")
    }
  }

  // MARK: - Keychain lookup

  private func loadPrivateKey(tag: Data) throws -> SecKey {
    // Returns a REFERENCE to the key (kSecReturnRef), never its raw bytes. A
    // data-returning keychain query is never issued, so private key material
    // cannot leave the keystore.
    let query: [String: Any] = [
      kSecClass as String: kSecClassKey,
      kSecAttrApplicationTag as String: tag,
      kSecAttrKeyType as String: keyType,
      kSecReturnRef as String: true,
    ]
    var item: CFTypeRef?
    let status = SecItemCopyMatching(query as CFDictionary, &item)
    guard status == errSecSuccess, let key = item else {
      throw CustodyException("handle not found")
    }
    // swiftlint:disable:next force_cast
    return (key as! SecKey)
  }

  // MARK: - Attestation & evidence

  private static func attest(
    privateKey: SecKey,
    profile: String,
    requireUserPresence: Bool
  ) throws -> [String: Any] {
    let attrs = SecKeyCopyAttributes(privateKey) as? [String: Any] ?? [:]
    let tokenId = attrs[kSecAttrTokenID as String] as? String
    let inEnclave = tokenId == (kSecAttrTokenIDSecureEnclave as String)

    let securityLevel: String
    let custodyClass: Int
    switch profile {
    case "ios-secure-enclave":
      securityLevel = "secure_enclave"
      // hardware_or_external only when residency is actually the enclave.
      custodyClass = inEnclave ? 3 : 2
    default:
      securityLevel = "keychain"
      custodyClass = 2 // os_protected
    }

    // presence: user_presence_required(4) when gated, else
    // unattended_after_first_unlock(2) for ThisDeviceOnly WhenUnlocked keys.
    let presenceMode = requireUserPresence ? 4 : 2

    return [
      "custodyClass": custodyClass,
      "presenceMode": presenceMode,
      "securityLevel": securityLevel,
      "profile": profile,
    ]
  }

  private static func providerEvidence(tag: Data, attestation: [String: Any]) -> Data {
    var evidence = Data()
    evidence.append(contentsOf: "flycockpit.ios-custody.v1".utf8)
    evidence.append(0)
    evidence.append(contentsOf: (attestation["securityLevel"] as? String ?? "").utf8)
    evidence.append(0)
    evidence.append(tag)
    return evidence
  }

  private static func publicCoordinates(of privateKey: SecKey) throws -> (Data, Data) {
    guard let publicKey = SecKeyCopyPublicKey(privateKey) else {
      throw CustodyException("failed to derive public key")
    }
    var error: Unmanaged<CFError>?
    // External representation of the PUBLIC key only: 0x04 || X(32) || Y(32).
    // The private key is Secure-Enclave / non-exportable; the OS refuses to
    // externalize it.
    guard let rep = SecKeyCopyExternalRepresentation(publicKey, &error) as Data? else {
      throw CustodyException("failed to export public key: \(describe(error))")
    }
    guard rep.count == 65, rep.first == 0x04 else {
      throw CustodyException("unexpected public key encoding")
    }
    return (rep.subdata(in: 1..<33), rep.subdata(in: 33..<65))
  }

  private static func profileForTag(_ tag: Data) throws -> (profile: String, requireUserPresence: Bool) {
    // TODO(native-platform): recover the key's REAL profile/presence from its
    // SecAccessControl / accessibility attributes (and the tag->profile mapping
    // in the durable store). Returning a hardcoded value here would silently
    // downgrade a user-presence-gated key to unattended, so we fail closed
    // instead of guessing.
    throw CustodyException(
      "unimplemented on this platform: native profile/presence recovery is not wired — failing closed"
    )
  }

  private static func describe(_ error: Unmanaged<CFError>?) -> String {
    guard let error = error?.takeRetainedValue() else { return "unknown error" }
    return CFErrorCopyDescription(error) as String? ?? "unknown error"
  }
}

private struct CustodyException: Error {
  let message: String
  init(_ message: String) { self.message = message }
}

// P-256 scalar helpers: DER→P1363 and low-S normalization.
private enum P256Scalar {
  // n = 0xffffffff00000000ffffffffffffffffbce6faada7179e84f3b9cac2fc632551
  static let order: [UInt8] = [
    0xff, 0xff, 0xff, 0xff, 0x00, 0x00, 0x00, 0x00, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff,
    0xbc, 0xe6, 0xfa, 0xad, 0xa7, 0x17, 0x9e, 0x84, 0xf3, 0xb9, 0xca, 0xc2, 0xfc, 0x63, 0x25, 0x51,
  ]
  // floor(n / 2)
  static let halfOrder: [UInt8] = [
    0x7f, 0xff, 0xff, 0xff, 0x80, 0x00, 0x00, 0x00, 0x7f, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff,
    0xde, 0x73, 0x7d, 0x56, 0xd3, 0x8b, 0xcf, 0x42, 0x79, 0xdc, 0xe5, 0x61, 0x7e, 0x31, 0x92, 0xa8,
  ]

  static func derToLowSP1363(_ der: Data) throws -> Data {
    let bytes = [UInt8](der)
    var i = 0
    func read() throws -> UInt8 {
      guard i < bytes.count else { throw CustodyException("corrupted signature") }
      defer { i += 1 }
      return bytes[i]
    }
    guard try read() == 0x30 else { throw CustodyException("corrupted signature") }
    _ = try read() // total length
    func readInt() throws -> [UInt8] {
      guard try read() == 0x02 else { throw CustodyException("corrupted signature") }
      let len = Int(try read())
      guard len > 0, i + len <= bytes.count else { throw CustodyException("corrupted signature") }
      var v = Array(bytes[i..<(i + len)])
      i += len
      while v.count > 1 && v.first == 0x00 { v.removeFirst() }
      while v.count < 32 { v.insert(0x00, at: 0) }
      guard v.count == 32 else { throw CustodyException("corrupted signature") }
      return v
    }
    let r = try readInt()
    var s = try readInt()

    if isZero(r) || isZero(s) || compare(r, order) >= 0 || compare(s, order) >= 0 {
      throw CustodyException("corrupted signature")
    }
    if compare(s, halfOrder) > 0 {
      s = subtract(order, s) // s := n - s
    }
    return Data(r + s)
  }

  private static func isZero(_ v: [UInt8]) -> Bool { v.allSatisfy { $0 == 0 } }

  private static func compare(_ a: [UInt8], _ b: [UInt8]) -> Int {
    for idx in 0..<32 {
      if a[idx] != b[idx] { return a[idx] < b[idx] ? -1 : 1 }
    }
    return 0
  }

  private static func subtract(_ a: [UInt8], _ b: [UInt8]) -> [UInt8] {
    var out = [UInt8](repeating: 0, count: 32)
    var borrow = 0
    for idx in stride(from: 31, through: 0, by: -1) {
      let diff = Int(a[idx]) - Int(b[idx]) - borrow
      if diff < 0 {
        out[idx] = UInt8(diff + 256)
        borrow = 1
      } else {
        out[idx] = UInt8(diff)
        borrow = 0
      }
    }
    return out
  }
}

File: apps/native/modules/remote-identity-custody/android/src/main/java/expo/modules/remoteidentitycustody/RemoteIdentityCustodyModule.kt
package expo.modules.remoteidentitycustody

import android.security.keystore.KeyGenParameterSpec
import android.security.keystore.KeyInfo
import android.security.keystore.KeyProperties
import expo.modules.kotlin.exception.CodedException
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import java.math.BigInteger
import java.security.KeyFactory
import java.security.KeyPairGenerator
import java.security.KeyStore
import java.security.Signature
import java.security.interfaces.ECPublicKey
import java.security.spec.ECGenParameterSpec

/**
 * RemoteIdentityCustody — Android durable P-256 custody.
 *
 * A durable handle is a non-exportable P-256 private key held in the Android
 * Keystore (StrongBox secure element when available, otherwise TEE, otherwise
 * software-backed). The `handleId` returned to JS is the key's random Keystore
 * alias. JS never receives private key bytes: the key is generated inside the
 * keystore and the app never reads its material back out.
 *
 * The alias→generation mapping (the app-layer monotonic high-water mark) is
 * persisted separately in app-private `SharedPreferences` by the app-layer
 * `NativeCustodyStore`. Diffie-Hellman key agreement is owned by the Noise
 * binding, never here — this module is P-256 signing only.
 */
class RemoteIdentityCustodyModule : Module() {
  private val keystoreProvider = "AndroidKeyStore"

  // Native platform backing (UNIMPLEMENTED — fails closed).
  //
  // TODO(native-platform): the durable native custody backing is NOT wired yet.
  // Until it is, every entry point fails closed rather than returning a
  // plausible-but-wrong result or relying on volatile state. What remains:
  //   1. a real durable store for the alias->generation high-water mark
  //      (app-private SharedPreferences), and
  //   2. recovery of a key's real profile/presence from its KeyInfo (see
  //      profileForAlias below).
  // The reference implementation of each method is retained below to document
  // the intended Android Keystore / StrongBox mechanism. This MUST be built and
  // verified on the Android CI leg — it cannot be compiled on the Linux gate box.
  // See NATIVE-PLATFORM-TODO.md.
  private fun requireNativeBackingWired() {
    throw CustodyException(
      "unimplemented on this platform: durable native custody backing is not wired — failing closed (see NATIVE-PLATFORM-TODO.md)",
    )
  }

  override fun definition() = ModuleDefinition {
    Name("RemoteIdentityCustody")

    AsyncFunction("generateP256") { handleId: ByteArray, profile: String, requireUserPresence: Boolean ->
      generate(handleId, profile, requireUserPresence)
    }

    AsyncFunction("signP256") { handleId: ByteArray, signingMessage: ByteArray ->
      sign(handleId, signingMessage)
    }

    AsyncFunction("publicKey") { handleId: ByteArray ->
      publicKeyReport(handleId)
    }

    AsyncFunction("rotateP256") { handleId: ByteArray, newHandleId: ByteArray ->
      rotate(handleId, newHandleId)
    }

    AsyncFunction("destroyGeneration") { handleId: ByteArray ->
      destroy(handleId)
    }
  }

  // MARK: - Generation

  private fun generate(handleId: ByteArray, profile: String, requireUserPresence: Boolean): Map<String, Any> {
    requireNativeBackingWired() // TODO(native-platform): fails closed until wired.
    // The caller-assigned 16-byte handle (the evidence subjectId) is the stable
    // Keystore alias source, so a durable write-ahead marker can name the key
    // before it exists.
    val alias = aliasFor(handleId)
    val useStrongBox = profile == "android-strongbox"

    val builder = KeyGenParameterSpec.Builder(alias, KeyProperties.PURPOSE_SIGN)
      .setDigests(KeyProperties.DIGEST_SHA256)
      .setAlgorithmParameterSpec(ECGenParameterSpec("secp256r1"))

    if (useStrongBox) {
      // Request hardware secure-element residency. If StrongBox is unavailable
      // the generator throws and no weaker key is silently substituted.
      builder.setIsStrongBoxBacked(true)
    }
    if (requireUserPresence) {
      // Presence-gated key: a signature requires a fresh user authentication.
      builder.setUserAuthenticationRequired(true)
    }

    val generator = KeyPairGenerator.getInstance(
      KeyProperties.KEY_ALGORITHM_EC,
      keystoreProvider,
    )
    generator.initialize(builder.build())
    val keyPair = generator.generateKeyPair()

    val (x, y) = publicCoordinates(keyPair.public as ECPublicKey)
    val attestation = attest(alias, profile, requireUserPresence)
    val evidence = providerEvidence(alias, attestation)

    return mapOf(
      "handleId" to handleId,
      "publicKey" to mapOf("x" to x, "y" to y),
      "attestation" to attestation,
      "providerEvidence" to evidence,
    )
  }

  // MARK: - Rotation

  private fun rotate(handleId: ByteArray, newHandleId: ByteArray): Map<String, Any> {
    requireNativeBackingWired() // TODO(native-platform): fails closed until wired.
    val alias = aliasFor(handleId)
    val store = KeyStore.getInstance(keystoreProvider).apply { load(null) }
    if (!store.containsAlias(alias)) {
      throw CustodyException("handle not found")
    }
    // Create a fresh key under the caller-assigned new alias; the old key is
    // retained until the app layer destroys it after publish.
    val existing = attest(alias, profileForAlias(alias), false)
    return generate(newHandleId, existing["profile"] as String, false)
  }

  // MARK: - Signing

  private fun sign(handleId: ByteArray, signingMessage: ByteArray): ByteArray {
    requireNativeBackingWired() // TODO(native-platform): fails closed until wired.
    val alias = aliasFor(handleId)
    val store = KeyStore.getInstance(keystoreProvider).apply { load(null) }
    val entry = store.getEntry(alias, null) as? KeyStore.PrivateKeyEntry
      ?: throw CustodyException("handle not found")

    // Message-based signing: the platform hashes with SHA-256 internally. The
    // private key is a keystore reference; its bytes never enter the process.
    val signer = Signature.getInstance("SHA256withECDSA")
    signer.initSign(entry.privateKey)
    signer.update(signingMessage)
    val der = signer.sign()

    // Convert DER (SEQUENCE { INTEGER r, INTEGER s }) to fixed 64-byte P1363 and
    // normalize to low-S. Zero / out-of-range components are corruption.
    return P256Scalar.derToLowSP1363(der)
  }

  // MARK: - Public key

  private fun publicKeyReport(handleId: ByteArray): Map<String, Any> {
    requireNativeBackingWired() // TODO(native-platform): fails closed until wired.
    val alias = aliasFor(handleId)
    val store = KeyStore.getInstance(keystoreProvider).apply { load(null) }
    val cert = store.getCertificate(alias) ?: throw CustodyException("handle not found")
    val (x, y) = publicCoordinates(cert.publicKey as ECPublicKey)
    val attestation = attest(alias, profileForAlias(alias), false)
    return mapOf("x" to x, "y" to y, "attestation" to attestation)
  }

  // MARK: - Destroy

  private fun destroy(handleId: ByteArray) {
    requireNativeBackingWired() // TODO(native-platform): fails closed until wired.
    val alias = aliasFor(handleId)
    val store = KeyStore.getInstance(keystoreProvider).apply { load(null) }
    if (!store.containsAlias(alias)) {
      throw CustodyException("handle not found")
    }
    store.deleteEntry(alias)
  }

  // MARK: - Attestation & evidence

  private fun attest(alias: String, profile: String, requireUserPresence: Boolean): Map<String, Any> {
    val store = KeyStore.getInstance(keystoreProvider).apply { load(null) }
    val securityLevel: String
    val custodyClass: Int

    val entry = store.getEntry(alias, null) as? KeyStore.PrivateKeyEntry
    if (entry != null) {
      val factory = KeyFactory.getInstance(entry.privateKey.algorithm, keystoreProvider)
      val info = factory.getKeySpec(entry.privateKey, KeyInfo::class.java)
      // Derive the real security level from KeyInfo, never from caller input.
      securityLevel = when {
        !info.isInsideSecureHardware -> "software"
        isStrongBox(info) -> "strongbox"
        else -> "tee"
      }
      // StrongBox is hardware_or_external(3); TEE and software are os_protected(2).
      custodyClass = if (securityLevel == "strongbox") 3 else 2
    } else {
      securityLevel = if (profile == "android-strongbox") "strongbox" else "tee"
      custodyClass = if (securityLevel == "strongbox") 3 else 2
    }

    // presence: user_presence_required(4) when auth-gated, else
    // unattended_unlocked_device(3) on Android.
    val presenceMode = if (requireUserPresence) 4 else 3

    return mapOf(
      "custodyClass" to custodyClass,
      "presenceMode" to presenceMode,
      "securityLevel" to securityLevel,
      "profile" to profile,
    )
  }

  private fun isStrongBox(info: KeyInfo): Boolean {
    return try {
      // API 31+: securityLevel == SECURITY_LEVEL_STRONGBOX.
      val method = KeyInfo::class.java.getMethod("getSecurityLevel")
      (method.invoke(info) as Int) == 2 // KeyProperties.SECURITY_LEVEL_STRONGBOX
    } catch (_: Throwable) {
      false
    }
  }

  private fun providerEvidence(alias: String, attestation: Map<String, Any>): ByteArray {
    val header = "flycockpit.android-custody.v1".toByteArray(Charsets.UTF_8)
    val level = (attestation["securityLevel"] as String).toByteArray(Charsets.UTF_8)
    val aliasBytes = alias.toByteArray(Charsets.UTF_8)
    return header + byteArrayOf(0) + level + byteArrayOf(0) + aliasBytes
  }

  private fun publicCoordinates(publicKey: ECPublicKey): Pair<ByteArray, ByteArray> {
    // Affine coordinates read directly from the public key — the private key is
    // never serialized or exported. Each coordinate is 32 bytes.
    val x = toFixed32(publicKey.w.affineX)
    val y = toFixed32(publicKey.w.affineY)
    return Pair(x, y)
  }

  private fun toFixed32(value: BigInteger): ByteArray {
    var bytes = value.toByteArray()
    if (bytes.size > 32) {
      bytes = bytes.copyOfRange(bytes.size - 32, bytes.size)
    }
    val out = ByteArray(32)
    System.arraycopy(bytes, 0, out, 32 - bytes.size, bytes.size)
    return out
  }

  private fun profileForAlias(alias: String): String {
    // TODO(native-platform): recover the key's REAL profile from its KeyInfo /
    // the durable alias->profile mapping. Defaulting to the strongest profile
    // here would MISREPORT a software/TEE key as StrongBox, so we fail closed
    // instead of guessing.
    throw CustodyException(
      "unimplemented on this platform: native profile/presence recovery is not wired — failing closed",
    )
  }

  private fun aliasFor(handleId: ByteArray): String {
    return "flycockpit-remote-identity-" + handleId.joinToString("") { "%02x".format(it) }
  }
}

private class CustodyException(message: String) :
  CodedException("ERR_REMOTE_IDENTITY_CUSTODY", message, null)

/** P-256 scalar helpers: DER→P1363 and low-S normalization. */
private object P256Scalar {
  // n = 0xffffffff00000000ffffffffffffffffbce6faada7179e84f3b9cac2fc632551
  private val order = BigInteger(
    "ffffffff00000000ffffffffffffffffbce6faada7179e84f3b9cac2fc632551",
    16,
  )
  private val halfOrder = order.shiftRight(1)

  fun derToLowSP1363(der: ByteArray): ByteArray {
    var i = 0
    fun read(): Int {
      if (i >= der.size) throw CustodyException("corrupted signature")
      return der[i++].toInt() and 0xff
    }
    if (read() != 0x30) throw CustodyException("corrupted signature")
    read() // total length
    fun readInt(): BigInteger {
      if (read() != 0x02) throw CustodyException("corrupted signature")
      val len = read()
      if (len <= 0 || i + len > der.size) throw CustodyException("corrupted signature")
      val v = der.copyOfRange(i, i + len)
      i += len
      return BigInteger(1, v)
    }
    val r = readInt()
    var s = readInt()

    if (r.signum() == 0 || s.signum() == 0 || r >= order || s >= order) {
      throw CustodyException("corrupted signature")
    }
    if (s > halfOrder) {
      s = order.subtract(s) // s := n - s
    }
    return toFixed32(r) + toFixed32(s)
  }

  private fun toFixed32(value: BigInteger): ByteArray {
    var bytes = value.toByteArray()
    if (bytes.size > 32) {
      bytes = bytes.copyOfRange(bytes.size - 32, bytes.size)
    }
    val out = ByteArray(32)
    System.arraycopy(bytes, 0, out, 32 - bytes.size, bytes.size)
    return out
  }
}
